pages/posts/modelling-inheritance.md

```diff
@@ -0,0 +1,233 @@
+---
+title: "Modeling Inheritance with JSON Schema"
+date: 2023-08-18
+tags:
+  - Questions
+  - Explorations
+type: Engineering
+cover: /img/posts/2023/modelling-inheritance/cover.webp
+authors:
+  - name: Greg Dennis
+    photo: /img/avatars/gregsdennis.webp
+    link: https://twitter.com/gregsdennis
+    byline: JSON Tooling Implementer, Specification & Tooling Architect @ Postman
+excerpt: Can it be done?  Well, sort of.
+---
+
+Probably the most common question that we get is, "How do I model an inheritance hierarchy in JSON Schema?"  And most commonly, our answer to that question is, "You don't."
+
+JSON Schema just isn't [designed](https://modern-json-schema.com/json-schema-is-a-constraint-system) for that.  It's a subtractive system, where more constraints means fewer matches, and data modelling tends to be additive, where more definition means more matches.  The systems are inherently incompatible.
+
+However, if we accept a few concessions, we might just be able to work something out.
+
+## Our Models
+
+To get started, we're going to attempt to model some computer peripherals.  In a strongly typed language, we may model this using a `Peripheral` base class that defines a number of properties (and typically functions) that are common to all peripherals.  Then, each device would be a subclass of this base class.
+
+For our purposes, we're just going to define the `name` property on the base class.  That is, every peripheral needs to have a name.
+
+_I'm going to use TypeScript for the code samples, but the concepts will apply to other languages as well._
+
+```ts
+abstract class Peripheral {
+  name: string;
+  // ...
+}
+```
+
+Now we can define other peripherals, `Mouse` and `Keyboard`, by inheriting from this base class.
+
+```ts
+class Mouse extends Peripheral {
+  buttonCount: number;
+  wheelCount: number;
+  trackingType: string; // ball / optical
+  // ...
+}
```

Comment:
Suggestion for more idiomatic TypeScript.

```suggestion
class Mouse extends Peripheral {
  buttonCount: number;
  wheelCount: number;
  trackingType: "ball" | "optical";
  // ...
}
```

Or, better but probably unnecessarily verbose for a blog post

```suggestion
type TrackingType = "ball" | "optical";

class Mouse extends Peripheral {
  buttonCount: number;
  wheelCount: number;
  trackingType: TrackingType;
  // ...
}
```